# Deep Nets using TF- Abstractions

[Prashant Brahmbhatt](www.github.com/hashbanger)

## Keras API

_____

### The Data

To compare these various abstractions we'll use a dataset easily available from the SciKit Learn library. The data is comprised of the results of a chemical analysis of wines grown in the same region in Italy by three different cultivators. There are thirteen different
measurements taken for different constituents found in the three types of wine. We will use the various TF Abstractions to classify the wine to one of the 3 possible labels.


In [16]:
from sklearn.datasets import load_wine

In [17]:
wine_data = load_wine()

In [18]:
type(wine_data)

sklearn.utils.Bunch

The sklearn Bunch is kind of a dictionary which contains the data as well as other information related to the data

In [19]:
wine_data.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

Taking a look at the description of the data.

In [20]:
print(wine_data['DESCR'])

Wine Data Database

Notes
-----
Data Set Characteristics:
    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- 1) Alcohol
 		- 2) Malic acid
 		- 3) Ash
		- 4) Alcalinity of ash  
 		- 5) Magnesium
		- 6) Total phenols
 		- 7) Flavanoids
 		- 8) Nonflavanoid phenols
 		- 9) Proanthocyanins
		- 10)Color intensity
 		- 11)Hue
 		- 12)OD280/OD315 of diluted wines
 		- 13)Proline
        	- class:
                - class_0
                - class_1
                - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash: 

____

Creating the features and the labels

In [21]:
feat_data = wine_data['data']

In [22]:
labels = wine_data['target']

### Splitting the dataset

In [23]:
from sklearn.model_selection import train_test_split

In [24]:
X_train, X_test, y_train, y_test = train_test_split(feat_data, labels, test_size = 0.3, random_state = 101)

### Scaling the data

In [25]:
from sklearn.preprocessing import MinMaxScaler

In [26]:
scaler = MinMaxScaler()

In [27]:
scaled_x_train = scaler.fit_transform(X_train)

In [28]:
scaled_x_test = scaler.fit_transform(X_test)

_______

## The abstraction

In [29]:
import keras
import tensorflow as tf

### Creating the model

In [30]:
from keras import Sequential

In [31]:
dnn_model = Sequential()

### Addding the layers

In [32]:
from keras import layers

In [33]:
dnn_model.add(layers.Dense(units= 13, input_dim = 13, activation= 'relu' ))

In [34]:
dnn_model.add(layers.Dense(units= 13, activation= 'relu' ))
dnn_model.add(layers.Dense(units= 13, activation= 'relu' ))              

In [35]:
dnn_model.add(layers.Dense(units= 3, activation= 'softmax' ))

### Compiling the model

In [36]:
from keras import optimizers, losses, metrics

There are several optimizers, losses to explore

In [37]:
#losses.
#optimizers.

In [38]:
dnn_model.compile(optimizer= 'adam', loss= 'sparse_categorical_crossentropy', metrics= ['accuracy'])

### Training the Model

In [39]:
dnn_model.fit(x = scaled_x_train, y = y_train, epochs= 100)

Epoch 1/100
124/124 [==============================] - 1s 4ms/step - loss: 1.0899 - acc: 0.4435
Epoch 2/100
124/124 [==============================] - 0s 88us/step - loss: 1.0806 - acc: 0.5242
Epoch 3/100
124/124 [==============================] - 0s 113us/step - loss: 1.0712 - acc: 0.5887
Epoch 4/100
124/124 [==============================] - 0s 129us/step - loss: 1.0621 - acc: 0.6129
Epoch 5/100
124/124 [==============================] - 0s 80us/step - loss: 1.0518 - acc: 0.6371
Epoch 6/100
124/124 [==============================] - 0s 97us/step - loss: 1.0404 - acc: 0.6452
Epoch 7/100
124/124 [==============================] - 0s 80us/step - loss: 1.0286 - acc: 0.6452
Epoch 8/100
124/124 [==============================] - 0s 88us/step - loss: 1.0151 - acc: 0.6452
Epoch 9/100
124/124 [==============================] - 0s 80us/step - loss: 0.9997 - acc: 0.6452
Epoch 10/100
124/124 [==============================] - 0s 121us/step - loss: 0.9828 - acc: 0.6532
Epoch 11/100
124/124 [=====

### Making the predcitions

In [43]:
preds_labels = dnn_model.predict_classes(x= scaled_x_test)

Evaluating using the confusion matrix

In [44]:
from sklearn.metrics import confusion_matrix, classification_report

In [45]:
confusion_matrix(y_true= y_test, y_pred= preds_labels)

array([[19,  0,  0],
       [ 0, 22,  0],
       [ 0,  1, 12]], dtype=int64)

In [46]:
print(classification_report(y_test, preds_labels))

             precision    recall  f1-score   support

          0       1.00      1.00      1.00        19
          1       0.96      1.00      0.98        22
          2       1.00      0.92      0.96        13

avg / total       0.98      0.98      0.98        54



So we got an accuracy floating around 98%

### de nada!